### ISSUE TITLE
numpy.test does not find all tests (migrated from Trac #1089)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#2642]
Original ticket http://projects.scipy.org/numpy/ticket/1089
Reported 2009-04-22 by atmention:cournape, assigned to unknown.

I noticed that nosetests runner runs a couple of more tests compared to numpy.test. Incidentally, most of the undetected tests fail

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2009-04-22

I see only distutils tests that are missing like this:

```
======================================================================
ERROR: Failure: ImportError (cannot import name fib2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../numpy/distutils/tests/f2py_ext/tests/test_fib2.py", line 3, in <module>
    from f2py_ext import fib2
ImportError: cannot import name fib2

======================================================================
ERROR: Failure: ImportError (cannot import name foo)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../numpy/distutils/tests/f2py_f90_ext/tests/test_foo.py", line 3, in <module>
    from f2py_f90_ext import foo
ImportError: cannot import name foo

======================================================================
ERROR: Failure: ImportError (cannot import name fib3)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../numpy/distutils/tests/gen_ext/tests/test_fib3.py", line 3, in <module>
    from gen_ext import fib3
ImportError: cannot import name fib3

======================================================================
ERROR: Failure: ImportError (No module named primes)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../numpy/distutils/tests/pyrex_ext/tests/test_primes.py", line 3, in <module>
    from pyrex_ext.primes import primes
ImportError: No module named primes

======================================================================
ERROR: Failure: ImportError (cannot import name example)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../numpy/distutils/tests/swig_ext/tests/test_example.py", line 3, in <module>
    from swig_ext import example
ImportError: cannot import name example

======================================================================
ERROR: Failure: ImportError (cannot import name example2)
----------------------------------------------------------------------
Traceback (most recent call last):
  File ".../numpy/distutils/tests/swig_ext/tests/test_example2.py", line 3, in <module>
    from swig_ext import example2
ImportError: cannot import name example2
```

**Comment by thouis** [comment]
Comment in Trac by atmention:cournape, 2009-06-01

There is also a significantly number of tests which are not run on windows (~1850 vs ~2050), we should find out why.

**Comment by thouis** [comment]
Comment in Trac by atmention:pv, 2010-10-10

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-03

This got worse for 1.5.1

Windows:
    Ran 2464 tests in 17.125s

```
OK (KNOWNFAIL=9, SKIP=5)
<nose.result.TextTestResult run=2464 errors=0 failures=0>
```

OS X:
    Ran 3006 tests in 15.240s

```
OK (KNOWNFAIL=4, SKIP=1)
Out[1]: <nose.result.TextTestResult run=3006 errors=0 failures=0>
```

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-03

Most of the tests that are not run are from a few files only. From f2py/tests/test_array_from_pyobj.py (~350 tests), this is unusual:
    for t in Type._type_names:
        exec '''\
    class test_%s_gen(unittest.TestCase,
                  _test_shared_memory
                  ):
        def setUp(self):
            self.type = Type(%r)
        array = lambda self,dims,intent,obj: Array(Type(%r),dims,intent,obj)
    ''' % (t,t,t)

lib/tests/test_format.py (~170 tests) has many tests defined in the file docstring (raw), which is odd.

The tests in core/tests/test_umath_complex.py  (~45 tests) are skipped because of non-conforming C99 implementation on Windows. So nothing wrong there.

I attach a log of all tests that are not run.

**Comment by thouis** [comment]
Attachment in Trac by atmention:rgommers, 2011-03-03: [tests-not-run-under-wine.txt](http://projects.scipy.org/numpy/attachment/ticket/1089/tests-not-run-under-wine.txt)

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-03

Actually, the test_format tests are also skipped on purpose:
    def test_memmap_roundtrip():
        # XXX: test crashes nose on windows. Fix this
        if not (sys.platform == 'win32' or sys.platform == 'cygwin'):

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-03

The f2py tests are skipped at least under Wine due to some scary-looking compiler detection code in f2py/tests/util.py. I attach the relevant function as a separate script, the output for me is:
    Ralf-Gommerss-iMac:~ rgommers$ wine .wine/drive_c/Python26/python.exe Code/util.py 
    (False, False, False)
    Ralf-Gommerss-iMac:~ rgommers$ python Code/util.py 
    (True, True, True)
I do have working compilers under Wine, but this code fails. I don't really feel like looking into that code more...

In conclusion, the OS X and Windows test counts differ only where they should except for the f2py thing.

**Comment by thouis** [comment]
Attachment in Trac by atmention:rgommers, 2011-03-03: [util.py](http://projects.scipy.org/numpy/attachment/ticket/1089/util.py)

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-03

np.test() gives me 3213 tests run with current master, np.test('full') 3227. Running nosetests gives me 3233. Comparing the difference between verbose outputs, after some cleaning due to a few run-specific printed messages like memory locations, the difference between nosetests and np.test() is exactly the 14 tests that are run extra for np.test('full'), plus the six tests that Pauli pasted above as failures.

Those six are probably not run because of the strange path: 
    .../numpy/distutils/tests/f2py_ext/tests/
The nose docs at http://somethingaboutorange.com/mrl/projects/nose/1.0.0/finding_tests.html say: "Be aware that plugins and command line options can change any of those rules." So this must be what's happening.

The distutils/tests/ folder has a setup.py in it which is weird, since that dir is not a module but in that setup.py subpackages are defined (with those six tests in them):
    Ralf-Gommerss-iMac:tests rgommers$ ls 
    f2py_ext        pyrex_ext       test_fcompiler_gnu.py   test_misc_util.pyc
    f2py_f90_ext        setup.py        test_fcompiler_gnu.pyc  test_npy_pkg_config.py
    gen_ext         swig_ext        test_misc_util.py   test_npy_pkg_config.pyc
    Ralf-Gommerss-iMac:tests rgommers$ cat setup.py 
    #!/usr/bin/env python
    def configuration(parent_package='',top_path=None):
        from numpy.distutils.misc_util import Configuration
        config = Configuration('testnumpydistutils',parent_package,top_path)
        config.add_subpackage('pyrex_ext')
        config.add_subpackage('f2py_ext')
        #config.add_subpackage('f2py_f90_ext')
        config.add_subpackage('swig_ext')
        config.add_subpackage('gen_ext')
        return config

```
if __name__ == "__main__":
    from numpy.distutils.core import setup
    setup(configuration=configuration)
```

So final question is: what are those tests doing there and should they be run by np.test('full'). I'm guessing no. In that case this ticket can be closed.

**Comment by thouis** [comment]
Comment in Trac by atmention:rgommers, 2011-03-26

Just came across the explanation of the last six tests. In {{{NoseTester.prepare_test_args}}}:
            # bypass these samples under distutils
            argv += ['--exclude','f2py_ext']
            argv += ['--exclude','f2py_f90_ext']
            argv += ['--exclude','gen_ext']
            argv += ['--exclude','pyrex_ext']
            argv += ['--exclude','swig_ext']
So case closed.
